src/Explorer/Panes/PublishNotebookPaneAdapter.tsx

```diff
@@ -1,156 +1,178 @@
-      onChangeTags: (newValue: string) => (this.tags = newValue),
-      onChangeImageSrc: (newValue: string) => (this.imageSrc = newValue),
-      onError: this.createFormErrorForLargeImageSelection,
-      clearFormError: this.clearFormError
-    };
-
-    return <PublishNotebookPaneComponent {...publishNotebookPaneProps} />;
-  };
-
-  private reset = (): void => {
-    this.isOpened = false;
-    this.isExecuting = false;
-    this.formError = undefined;
-    this.formErrorDetail = undefined;
-    this.name = undefined;
-    this.author = undefined;
-    this.content = undefined;
-  };
-}
+import ko from "knockout";
+import * as React from "react";
+import { ReactAdapter } from "../../Bindings/ReactBindingHandler";
+import * as Logger from "../../Common/Logger";
+import Explorer from "../Explorer";
+import { JunoClient } from "../../Juno/JunoClient";
+import { NotificationConsoleUtils } from "../../Utils/NotificationConsoleUtils";
+import { ConsoleDataType } from "../Menus/NotificationConsole/NotificationConsoleComponent";
+import { GenericRightPaneComponent, GenericRightPaneProps } from "./GenericRightPaneComponent";
+import { PublishNotebookPaneComponent, PublishNotebookPaneProps } from "./PublishNotebookPaneComponent";
+import { ImmutableNotebook } from "@nteract/commutable/src";
+import { toJS } from "@nteract/commutable";
+
+export class PublishNotebookPaneAdapter implements ReactAdapter {
+  parameters: ko.Observable<number>;
+  private isOpened: boolean;
+  private isExecuting: boolean;
+  private formError: string;
+  private formErrorDetail: string;
+
+  private name: string;
+  private author: string;
+  private content: string;
+  private description: string;
+  private tags: string;
+  private imageSrc: string;
+  private notebookObject: ImmutableNotebook;
+  private parentDomElement: HTMLElement;
+
+  constructor(private container: Explorer, private junoClient: JunoClient) {
+    this.parameters = ko.observable(Date.now());
+    this.reset();
+    this.triggerRender();
+  }
+
+  public renderComponent(): JSX.Element {
+    if (!this.isOpened) {
+      return undefined;
+    }
+
+    const props: GenericRightPaneProps = {
+      container: this.container,
+      content: this.createContent(),
+      formError: this.formError,
+      formErrorDetail: this.formErrorDetail,
+      id: "publishnotebookpane",
+      isExecuting: this.isExecuting,
+      title: "Publish to gallery",
+      submitButtonText: "Publish",
+      onClose: () => this.close(),
+      onSubmit: () => this.submit()
+    };
+
+    return <GenericRightPaneComponent {...props} />;
+  }
+
+  public triggerRender(): void {
+    window.requestAnimationFrame(() => this.parameters(Date.now()));
+  }
+
+  public open(
+    name: string,
+    author: string,
+    notebookContent: string | ImmutableNotebook,
+    parentDomElement: HTMLElement
+  ): void {
+    this.name = name;
+    this.author = author;
+    if (typeof notebookContent === "string") {

```

Comment:
I'm a little bit uneasy with the content being either string of object, because now every caller has to check the type, but I can't think of a better solution (returning always string by stringify'ing and parsing again seems inefficient indeed).

Reply:
yes I considered the same, but seemed inefficient. also, publish is the only action that calls the getContent() in the notebookComponentBootstrapper, where this notebookContent is defined as string | ImmutableNotebook. so we are good.